mint-client/src/bin/clientd/main.rs

```diff
@@ -0,0 +1,297 @@
+use minimint::config::load_from_file;
+use mint_client::clients::user::APIResponse;
+use mint_client::rpc::{Request, Response, Router, Shared};
+use mint_client::{ClientAndGatewayConfig, UserClient};
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+use structopt::StructOpt;
+use tracing_subscriber::EnvFilter;
+
+#[derive(Clone)]
+pub struct State {
+    router: Arc<Router>,
+    shared: Arc<Shared>,
+}
+
+#[derive(StructOpt)]
+struct Options {
+    workdir: PathBuf,
+}
+
+#[tokio::main]
+async fn main() -> tide::Result<()> {
+    tracing_subscriber::fmt()
+        .with_env_filter(
+            EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("info")),
+        )
+        .with_writer(std::io::stderr)
+        .init();
+    let opts: Options = StructOpt::from_args();
+    let cfg_path = opts.workdir.join("client.json");
+    let db_path = opts.workdir.join("client.db");
+    let cfg: ClientAndGatewayConfig = load_from_file(&cfg_path);
+    let db = sled::open(&db_path)
+        .unwrap()
+        .open_tree("mint-client")
+        .unwrap();
+
+    let client = UserClient::new(cfg.client, Box::new(db), Default::default());
+    let router = Router::new().add_handler("info", info);
+    let shared = Shared {
+        client: Arc::new(client),
+        gateway: Arc::new(cfg.gateway.clone()),
+        events: Arc::new(Mutex::new(Vec::new())),
+    };
+    let state = State {
+        router: Arc::new(router),
+        shared: Arc::new(shared),
+    };
+    let mut app = tide::with_state(state);
+
+    /*app.at("/info").post(info);
+    app.at("/pegin_address").post(pegin_address);
+    app.at("pegin").post(pegin);
+    app.at("pegout").post(pegout);
+    app.at("/spend").post(spend);
+    app.at("/lnpay").post(ln_pay);
+    app.at("/reissue").post(reissue);
+    app.at("/reissue_validate").post(reissue_validate);
+    app.at("/pending").post(pending);
+    app.at("/events").post(events);
+     */
+    app.at("/rpc")
+        .post(|mut req: tide::Request<State>| async move {
+            let router = Arc::clone(&req.state().router);
+            let shared = Arc::clone(&req.state().shared);
```

Comment:
why not just `req.state().router.clone()`?

Reply:
Now that I thought about it I am a bit more confused: I need Arc for the client for example because the different endpoints need to own the SAME client but it shouldn't be dropped when one endpoint drops it. With Router it doesn't matter if you just have a clone and then drop it since it won't (and shouldn't) change based on what other /rpc do with it so there is no use for Arc ?

Reply:
Idk what's more efficient: cloning the `HashMap` in `Router` or cloning an `Arc`. One thing to consider though: why not put the router object into the `Shared` struct? That would make once `Arc::clone` redundant.